File: solutions/560-subarray-sum-equals-k/solution.go
package main

import "fmt"

func subarraySum(nums []int, k int) int { // 暴力枚举
	count := 0
	for start := 0; start < len(nums); start++ {
		sum := 0
		for end := start; end >= 0; end-- {
			sum += nums[end]
			if sum == k {
				count++
			}
		}
	}
	return count
}

func subarraySum2(nums []int, k int) int {
	// 前缀和 + 哈希表
	count := 0
	sum := 0
	prefixSum := make(map[int]int)
	prefixSum[0] = 1
	for _, num := range nums {
		sum += num
		count += prefixSum[sum-k]
		prefixSum[sum]++
	}
	return count
}

func main() {
	got := subarraySum([]int{1, 1, 1}, 2)
	status := "PASS"
	if got != 2 {
		status = "FAIL"
	}
	fmt.Printf("%s | count=%d\n", status, got)
}

File: solutions/560-subarray-sum-equals-k/solution.py
from typing import List


class Solution:
    def subarray_sum(self, nums: List[int], k: int) -> int:
        count = 0
        for start in range(len(nums)):
            s = 0
            for end in range(start, -1, -1):
                s += nums[end]
                if s == k:
                    count += 1
        return count

    def subarray_sum2(self, nums: List[int], k: int) -> int:
        count = 0
        s = 0
        prefix_sum = {0: 1}
        for num in nums:
            s += num
            count += prefix_sum.get(s - k, 0)
            prefix_sum[s] = prefix_sum.get(s, 0) + 1
        return count


if __name__ == "__main__":
    sol = Solution()
    got = sol.subarray_sum([1, 1, 1], 2)
    status = "PASS" if got == 2 else "FAIL"
    print(f"{status} | => {got}")
